bam_fastaref.c

```diff
@@ -0,0 +1,306 @@
+#include <fcntl.h>
+#include <unistd.h>
+
+#include "htslib/sam.h"
+#include "htslib/ref.h"
+#include "htslib/kstring.h"
+#include "htslib/khash.h"
+#include "htslib/hfile.h"
+#include "htslib/cram.h"
+#include "sam_header.h"
+
+KHASH_MAP_INIT_STR(s2s, char*)
+
+typedef struct{
+    int maxLineLength;
+    char (*outputKeys)[3];
+    int numOutputKeys;
+    char* samFileName;
+    char* outFileName; // NULL if this is to output to stdout
+} fastaref_options_t;
+
+/**
+ * Replacement in C99 for strsep
+*/
+char* _strsep(char **stringp, const char *delim){
+    if(*stringp == NULL) return NULL;
+    char* foundChar = strpbrk(*stringp, delim);
+
+    if(foundChar == NULL){
+        char* token = *stringp;
+        *stringp = NULL;
+        return token;
+    }
+
+    *foundChar = 0;
+
+    char* token = *stringp;
+    *stringp = foundChar + 1;
+    return token;
+}
+
+/**
+ * Given an SQ line, returns a hash map, containing it's fields.
+ * Note: this modifies the text in the sqLine parameter
+ */
+kh_s2s_t* parseSQLine(char* sqLine){
+    char* parsePosition = sqLine + 4; // 4 = len("@SQ\t")
+    char* field;
+    kh_s2s_t* sqLineDict = kh_init(s2s);
+
+    while((field = _strsep(&parsePosition, "\t")) != NULL){
+        if(strlen(field) < 4){
+            // cannot have a field of length < 4, we need to have "FD:" present
+            // where FD is the field name
+            print_error("fastaref", "failed parsing SQ line in header");
+            return NULL;
+        }
+
+        char* key = field;
+        key[2] = '\0';
+
+        char* value = field + 3;
+
+        int ret;
+        int valuePointer = kh_put(s2s, sqLineDict, key, &ret);
+        if(ret == -1){
+            print_error("fastaref", "failed parsing SQ line in header");
+            return NULL;
+        }
+        else if(ret == 0){
+            print_error("fastaref", "duplicate field name in SQ line");
+            return NULL;
+        }
+        kh_value(sqLineDict, valuePointer) = value;
+    }
+
+    return sqLineDict;
+}
+
+/**
+ * Generates a fasta file according to command line options specified in options
+ *
+ * @returns 0 if successful
+ */
+int generateFastaFile(fastaref_options_t *options){
+    int writeBufferLength = options->maxLineLength;
+    char* writeBuffer = malloc(writeBufferLength);
+    if(writeBuffer == NULL) return -1;
+    FILE* outFile = NULL;
+    int returnCode = 0;
+    bam_hdr_t* header = NULL;
+    hFILE* ref = NULL;
+    kh_s2s_t* sqLineDict = NULL;
+
+    samFile* inFile = sam_open(options->samFileName, "r");
+    if (inFile == NULL) {
+        print_error_errno("fastaref", "failed to open \"%s\"", options->samFileName);
+        returnCode = 1;
+        goto cleanup;
+    }
+
+    header = sam_hdr_read(inFile);
+    if (header == NULL) {
+        print_error("fastaref", "failed to read header for \"%s\"", options->samFileName);
+        returnCode = 1;
+        goto cleanup;
+    }
+
+    char* readPosition = header->text;
```

Comment:
Your parsing logic operates directly on the `text` string of the header structure. Not only does `_strsep` change the original string, but it could make the initial text pointer **NULL**, if it doesn't find a delimiter. You would then leak memory. 
I suggest making a copy of the header text (with `strdup`) and work on the copy. You should also save the initial pointer of the copy and use it at the end, when freeing the memory allocated to the copy. 
Thus, you decouple your logic from the header layer.